File: ru.kmorozov.library.data.server/src/main/java/ru/kmorozov/library/data/loader/impl/BaseLoader.java
package ru.kmorozov.library.data.loader.impl;

import ru.kmorozov.gbd.logger.Logger;
import ru.kmorozov.library.data.loader.ILoader;
import ru.kmorozov.library.data.model.book.Storage;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.ArrayList;
import java.util.Collection;

/**
 * Created by sbt-morozov-kv on 14.03.2017.
 */
public abstract class BaseLoader implements ILoader, Runnable {

    private static final Logger logger = Logger.getLogger(BaseLoader.class);
    protected Collection<Object> links = new ArrayList<>();
    protected volatile LoaderExecutor.State state = LoaderExecutor.State.STOPPED;

    @Override
    public void run() {
        try {
            this.load();
        } catch (final IOException | UncheckedIOException e) {
            state = LoaderExecutor.State.STOPPED;

            e.printStackTrace();
        }
    }

    void setState(LoaderExecutor.State state) {
        this.state = state;
    }

    LoaderExecutor.State getState() {
        return this.state;
    }

    public boolean isStopped() {
        return LoaderExecutor.State.STOPPED == this.state;
    }

    public abstract Storage refresh(Storage storage);
}

File: ru.kmorozov.library.data.server/src/main/java/ru/kmorozov/library/data/loader/impl/BaseLoader.kt
package ru.kmorozov.library.data.loader.impl

import ru.kmorozov.gbd.core.config.GBDOptions
import ru.kmorozov.gbd.logger.Logger
import ru.kmorozov.library.data.loader.ILoader
import ru.kmorozov.library.data.loader.impl.LoaderExecutor.State
import ru.kmorozov.library.data.model.book.Storage
import java.io.IOException
import java.io.UncheckedIOException

/**
 * Created by sbt-morozov-kv on 14.03.2017.
 */
abstract class BaseLoader : ILoader, Runnable {
    protected var links: MutableList<Any> = ArrayList()
    @Volatile
    internal var state = State.STOPPED

    val isStopped: Boolean
        get() = State.STOPPED == state

    override fun run() {
        try {
            load()
        } catch (e: IOException) {
            this.state = State.STOPPED

            e.printStackTrace()
        } catch (e: UncheckedIOException) {
            this.state = State.STOPPED
            e.printStackTrace()
        }

    }

    abstract fun refresh(storage: Storage): Storage

    companion object {

        private val logger = Logger.getLogger(GBDOptions.debugEnabled, BaseLoader::class.java)
    }
}
